core/src/main/scala/org/apache/spark/scheduler/DecommissionTracker.scala

```diff
@@ -0,0 +1,405 @@
+    }
+  }
+
+  /*
+   * Is the node decommissioning i.e from driver point of
+   * view the node is candidate for decommissioning.
+   * Not necessarily decommissioned or terminated
+   */
+  def isNodeDecommissioning(hostname: String): Boolean = synchronized {
+    decommissionHostnameMap.contains(hostname)
+  }
+
+  /**
+   * visible only for Unit Test
+   */
+  def getDecommissionedNodeState(hostname:
+                                 String): Option[NodeDecommissionState.Value] = synchronized {
+    decommissionHostnameMap.get(hostname) match {
+      case Some(info) => Some(info.state)
+      case _ => None
+    }
+  }
+
+  def addNodeToDecommission(hostname: String, terminationTimeMs: Long,
+                            reason: NodeDecommissionReason): Unit = synchronized {
+
+    val df: SimpleDateFormat = new SimpleDateFormat("YY/MM/dd HH:mm:ss")
+    val tDateTime = df.format(terminationTimeMs)
+    val curTimeMs = clock.getTimeMillis()
+
+    if (terminationTimeMs <= curTimeMs) {
+      // Ignoring the decommission request if termination
+      // time is less than or same as the current time
+      logWarning(s"Ignoring decommissioning request for host $hostname as" +
+        s" terminationTimeMs ${terminationTimeMs} is less" +
+        s" than current time ${curTimeMs}")
+      return
+    }
+
+    // Override decommissionHostnameMap in case termination time is less than
+    // existing the terminationTime in decommissionHostnameMap.
+    if (decommissionHostnameMap.contains(hostname)) {
+      val nodeDecommissionInfo = decommissionHostnameMap(hostname)
+      // There will be no duplicate entry of terminationTimeMs in decommissionHostnameMap
+      // since the terminationTime is updated only when it is less than the existing termination
+      // time in decommissionHostnameMap
+      if (decommissionHostnameMap(hostname).terminationTime <= terminationTimeMs) {
+        logDebug(
+          s"""Ignoring decommissioning """ +
+            s""" request : {"node":"$hostname","reason":"${reason.message}",terminationTime"""" +
+            s""":"$tDateTime"} current : {"node":"$hostname",$nodeDecommissionInfo}""")
+        return
+      } else {
+        logInfo(s"Updating the termination time to :${terminationTimeMs} in " +
+          s"decommission tracker for the hostname ${hostname}")
+      }
+    }
+
+    val delay = terminationTimeMs - curTimeMs
+
+    var executorDecommissionTimeMs = terminationTimeMs
+    var shuffleDataDecommissionTimeMs = terminationTimeMs
+
+    // if delay is less than a minDecommissionTime than decommission immediately
+    if (terminationTimeMs - curTimeMs < minDecommissionTime * 1000) {
+      executorDecommissionTimeMs = curTimeMs
+      // Added the delay of 1 second in case of delay is less than a minute
+      // Since we want executor to be decommissioned first
+      // than after that shuffleDataDecommission
+      shuffleDataDecommissionTimeMs = curTimeMs + 1000
```

Comment:
This is the life cycle of the nodes which is decommissioned
DECOMMISSIONING -> EXECUTOR_DECOMMISSIONED -> SHUFFLEDATA_DECOMMISSIONED -> TERMINATED.

**Now the query is what happen if SHUFFLEDATA_DECOMMISSIONED  happens before or at same time of the EXECUTOR_DECOMMISSIONED** - If we make SHUFFLEDATA_DECOMMISSIONED before the EXECUTOR_DECOMMISSIONED than there is a chance that some of the task will erred out due to fetch failed exception since we have already done SHUFFLEDATA_DECOMMISSIONED here. 

Also I have explained this here in this comment https://github.com/apache/spark/pull/27636#issuecomment-593338415

Reply:
perhaps that explanation needs to be added to description of lira as a design aspect. Also goes back to if need to happen in sequence then perhaps we should do in sequence in decommission tracker rather then in separate threads?

Reply:
sure will update jira with explanation given

**Also goes back to if need to happen in sequence then perhaps we should do in sequence in decommission tracker rather then in separate threads?** - I have explained above comment https://github.com/apache/spark/pull/27636/files#r393149928